File: iosApp/iosApp/Data/Storage/StorageManager.swift
import Foundation
import OSLog

/// User state matching Android's UserState sealed class
enum UserState: Equatable {
    case loadingUser
    case noUser
    case connectedWallet(address: String)
    case connectedUser(user: User)
    case userError(message: String)
}

/// Storage Manager matching Android's StorageManager.kt
/// Uses UserDefaults with Combine for reactive state management
class StorageManager: ObservableObject {
    // MARK: - Properties
    
    @Published private(set) var userState: UserState = .loadingUser
    
    private let userDefaults: UserDefaults
    private let jsonEncoder = JSONEncoder()
    private let jsonDecoder = JSONDecoder()
    
    private enum Keys {
        static let userProfile = "user_profile"
    }
    
    // MARK: - Initialization
    
    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
        loadUserState()
    }
    
    // MARK: - Public Methods
    
    /// Saves the user profile and updates the state
    func saveUserProfile(_ user: User) {
        Logger.storage.debug("StorageManager: Saving user profile (id: \(user.id), address: \(user.walletAddress, privacy: .private))")
        do {
            let data = try jsonEncoder.encode(user)
            userDefaults.set(data, forKey: Keys.userProfile)
            userState = .connectedUser(user: user)
            Logger.storage.info("StorageManager: User profile saved, state updated to .connectedUser")
        } catch {
            Logger.storage.error("StorageManager: Failed to save user profile: \(error)")
            userState = .userError(message: "Failed to save user profile: \(error.localizedDescription)")
        }
    }
    
    /// Saves the wallet address and updates the state
    func saveWalletAddress(_ address: String) {
        Logger.storage.debug("StorageManager: Saving wallet address: \(address, privacy: .private)")
        userState = .connectedWallet(address: address)
        Logger.storage.info("StorageManager: State updated to .connectedWallet")
    }

    /// Clears the user profile and updates the state
    func clearUserProfile() {
        userDefaults.removeObject(forKey: Keys.userProfile)
        userState = .noUser
        Logger.storage.info("StorageManager: Cleared user profile")
    }
    
    /// Returns the stored user if available
    func getStoredUser() -> User? {
        if case let .connectedUser(user) = userState {
            return user
        }
        return nil
    }
    
    /// Returns the stored wallet address if available
    func getStoredWallet() -> String? {
        switch userState {
        case .connectedUser(let user):
            return user.walletAddress
        case .connectedWallet(let address):
            return address
        case .noUser, .loadingUser, .userError:
            return nil
        }
    }
    
    /// Checks if a user profile exists
    func hasUserProfile() -> Bool {
        if case .connectedUser = userState {
            return true
        }
        return false
    }
    
    // MARK: - Private Methods
    
    private func loadUserState() {
        if let userData = userDefaults.data(forKey: Keys.userProfile),
           let user = try? jsonDecoder.decode(User.self, from: userData) {
            userState = .connectedUser(user: user)
            Logger.storage.info("StorageManager: Loaded user profile: \(user.id)")
        } else {
            userState = .noUser
            Logger.storage.debug("StorageManager: No user or wallet found in storage")
        }
    }
}

File: androidApp/src/main/java/org/idos/app/data/StorageManager.kt
package org.idos.app.data

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import org.idos.app.data.model.UserModel
import org.kethereum.model.Address
import timber.log.Timber

sealed class UserState

object LoadingUser : UserState()

object NoUser : UserState()

data class ConnectedWallet(
    val address: Address,
) : UserState()

data class ConnectedUser(
    val user: UserModel,
) : UserState()

data class UserError(
    val message: String,
) : UserState()

/**
 * StorageManager handles user data and app state using SharedPreferences with StateFlow.
 * This provides reactive state management for user authentication state.
 */
class StorageManager(
    private val context: Context,
    private val json: Json,
) {
    // SharedPreferences for user data
    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    // User state with Loading/Value/Error pattern
    private val _userState = MutableStateFlow<UserState>(LoadingUser)
    val userState: StateFlow<UserState> = _userState.asStateFlow()

    /**
     * Initialize storage manager by loading user state from SharedPreferences.
     * Safe to call multiple times - will reload state from disk.
     *
     * @param onUserLoaded Optional callback to validate/initialize user before emitting state.
     *                     Return true to emit ConnectedUser, false to skip emission.
     */
    suspend fun initialize(onUserLoaded: (suspend (UserModel) -> Boolean)? = null) {
        loadUserState(onUserLoaded)
    }

    /**
     * Load user state from SharedPreferences on background thread
     */
    private suspend fun loadUserState(onUserLoaded: (suspend (UserModel) -> Boolean)?) =
        withContext(Dispatchers.IO) {
            try {
                val userJson = prefs.getString(KEY_USER_PROFILE, null)
                if (userJson != null) {
                    val userModel = json.decodeFromString<UserModel>(userJson)

                    // Allow caller to validate/initialize before emitting state
                    val shouldEmit = onUserLoaded?.invoke(userModel) ?: true

                    if (shouldEmit) {
                        _userState.value = ConnectedUser(userModel)
                        Timber.d("Loaded user profile: ${userModel.id}")
                    }
                } else {
                    _userState.value = NoUser
                    Timber.d("No user profile found in storage")
                }
            } catch (e: SerializationException) {
                // there is an issue with serializer, we might have updated the model
                Timber.w(e, "Failed to load user profile, clearing and logging out")
                clearUserProfile()
            } catch (e: Exception) {
                Timber.e(e, "Failed to load user profile")
                _userState.value = UserError("Failed to load user profile: ${e.message}")
            }
        }

    /**
     * Save user profile and update state on background thread
     */
    suspend fun saveUserProfile(userModel: UserModel) =
        withContext(Dispatchers.IO) {
            try {
                val userJson = json.encodeToString(userModel)
                prefs.edit {
                    putString(KEY_USER_PROFILE, userJson)
                }
                _userState.value = ConnectedUser(userModel)
                Timber.d("Saved user profile: ${userModel.id}")
            } catch (e: Exception) {
                Timber.e(e, "Failed to save user profile")
                _userState.value = UserError("Failed to save user profile: ${e.message}")
            }
        }

    /**
     * Save wallet address and update state
     * This doesn't touch disk, just updates in-memory state
     */
    fun saveWalletAddress(address: Address) {
        _userState.value = ConnectedWallet(address)
    }

    /**
     * Clear user profile and update state on background thread
     */
    suspend fun clearUserProfile() =
        withContext(Dispatchers.IO) {
            try {
                prefs.edit {
                    remove(KEY_USER_PROFILE)
                }
                _userState.value = NoUser
                Timber.d("Cleared user profile")
            } catch (e: Exception) {
                Timber.e(e, "Failed to clear user profile")
                _userState.value = UserError("Failed to clear user profile: ${e.message}")
            }
        }

    /**
     * Utility methods
     */
    fun hasUserProfile(): Boolean = _userState.value is ConnectedUser

    fun getStoredUser(): UserModel? =
        when (val state = _userState.value) {
            is ConnectedUser -> state.user
            else -> null
        }

    fun getStoredWallet(): Address? =
        when (val state = _userState.value) {
            is ConnectedUser -> state.user.walletAddress
            is ConnectedWallet -> state.address
            else -> null
        }

    companion object {
        private const val PREFS_NAME = "idos_storage_manager"
        private const val KEY_USER_PROFILE = "user_profile"
    }
}
